staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root/discovery_test.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package root
+
+import (
+	"net/http"
+	"net/http/httptest"
+
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	apidiscoveryv2beta1 "k8s.io/api/apidiscovery/v2beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	runtimeserializer "k8s.io/apimachinery/pkg/runtime/serializer"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	k8sscheme "k8s.io/client-go/kubernetes/scheme"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+)
+
+var scheme = runtime.NewScheme()
+var codecs = runtimeserializer.NewCodecFactory(scheme)
+var serializer runtime.NegotiatedSerializer
+
+const discoveryPath = "/apis"
+const jsonAccept = "application/json"
+const aggregatedAccept = ";g=apidiscovery.k8s.io;v=v2beta1;as=APIGroupDiscoveryList"
+
+func init() {
+	// Add all builtin types to scheme
+	k8sscheme.AddToScheme(scheme)
+	apidiscoveryv2beta1.AddToScheme(scheme)
+	info, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), runtime.ContentTypeJSON)
+	if !ok {
+		panic("failed to create serializer info")
+	}
+	serializer = runtime.NewSimpleNegotiatedSerializer(info)
+}
+
+func fetchPath(handler http.Handler, path, accept string) string {
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest("GET", discoveryPath, nil)
+
+	// Ask for JSON response
+	req.Header.Set("Accept", accept)
+
+	handler.ServeHTTP(w, req)
+	return string(w.Body.Bytes())
+}
+
+type fakeHTTPHandler struct {
+	data string
+}
+
+func (f fakeHTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	io.WriteString(resp, f.data)
+}
+
+func TestAggregationEnabled(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.AggregatedDiscoveryEndpoint, true)()
+
+	unaggregated := fakeHTTPHandler{data: "unaggregated"}
+	aggregated := fakeHTTPHandler{data: "aggregated"}
+	rootHandler := NewRootAPIsHandler(serializer, unaggregated, aggregated)
+
+	testCases := []struct {
+		accept   string
+		expected string
+	}{
+		{
+			accept:   jsonAccept + aggregatedAccept,
+			expected: "aggregated",
+		}, {
+			accept:   jsonAccept,
+			expected: "unaggregated",
+		}, {
+			// Server should return the first accepted type
+			accept:   jsonAccept + aggregatedAccept + "," + jsonAccept,
+			expected: "aggregated",
+		},
```

Comment:
Maybe add a test with something that "doesn't exist" (i.e. a misconstructed accept entry).